### ISSUE TITLE
A Theoretically Grounded Application of Dropout in RNN (2016)

### ISSUE THREAD

**Issue by esvhd** [esvhd/notes#1]
[paper](https://arxiv.org/abs/1512.05287) by Gal & Ghahramani, 2016.

Lua code available [here](https://github.com/yaringal/BayesianRNN)

Introduces **Variational** LSTM: the same dropout mask is used for **all** time steps in the recurrent steps. 

![dropout](https://user-images.githubusercontent.com/7330385/39837953-ca8e65ac-53cf-11e8-9781-d10e03bfc09b.png)

## Formulation

* **untied-weights** LSTM, formula (5)
* **tied-weights** LSTM, formula (6), faster forward pass, slightly worse results.

For **untied-weights**, different dropout masks could be used for different **gates**. 

![lstm](https://user-images.githubusercontent.com/7330385/39838561-a03273c8-53d1-11e8-926a-41bad2418ef4.png)

## Monte Carlo (MC) Dropout: 

Obtained by performing dropout at **test time** 1000 times, and averaging the model outputs following equation (4) in paper.

![predict](https://user-images.githubusercontent.com/7330385/39839416-ef242952-53d3-11e8-819c-0939dc20b965.png)

 ## Experiments / Results

* Larger model with dropout performed better than smaller model with dropout.
* Untied-weights + MC dropout performed better than tied-weights + MC dropout.

![results](https://user-images.githubusercontent.com/7330385/39839850-74f39c7e-53d5-11e8-93f5-1df6c6719d38.png)

**Comment by esvhd** [comment]
Found some nice notes here on this topic, and compares `tensorflow` and `pytorch` implementations.

* [I](https://becominghuman.ai/learning-note-dropout-in-recurrent-networks-part-1-57a9c19a2307) explains for the maths, also talks about the implementations in `tensorflow`/`keras` and `pytorch`.

* [II](https://towardsdatascience.com/learning-note-dropout-in-recurrent-networks-part-2-f209222481f8) Compares with Merity AWD-LSTM-LM, which allows different mask for different gates for tied-weight LSTM. Detailed notes on `keras` LSTM.

* [III](https://towardsdatascience.com/learning-note-dropout-in-recurrent-networks-part-3-1b161d030cd4)
